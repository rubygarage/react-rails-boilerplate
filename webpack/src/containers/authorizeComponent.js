import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { getCurrentUser } from 'selectors/user';

function authorizeComponent(
  WrappedComponent,
  fallbackLocation = '/404',
  authorizingFunction = (currentUser, match) => (currentUser.id === match.params.id),
) {
  const mapStateToProps = state => ({ currentUser: getCurrentUser(state) });
  return connect(mapStateToProps)((props) => {
    if (!authorizingFunction(props.currentUser, props.match)) {
      return <Redirect to={fallbackLocation} />;
    }
    return <WrappedComponent {...props} />;
  });
}

export default authorizeComponent;
