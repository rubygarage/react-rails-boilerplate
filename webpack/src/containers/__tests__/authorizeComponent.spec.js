import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import authorizeComponent from '../authorizeComponent';

describe('authorizeComponent HoC', () => {
  const defaultProps = {
    match: {
      params: {
        id: '1',
      },
    },
  };
  const MockComponent = () => <div>Mock Component</div>;
  const WrappedComponent = authorizeComponent(MockComponent);

  describe('user is authorized', () => {
    const store = configureStore()({
      entities: {
        users: {
          1: { id: '1', username: 'name' },
          2: { id: '2', username: 'name2' },
        },
      },
      signin: { currentUser: { users: ['1'] } },
    });

    it('renders given component', () => {
      const authenticatedContainer = mount(<WrappedComponent {...defaultProps} store={store} />);
      expect(authenticatedContainer).toMatchSnapshot();
    });

    it('allows to use custom authorization function', () => {
      const WithCustomFunction = authorizeComponent(
        MockComponent,
        undefined,
        (currentUser, match) => (currentUser.username === 'name' && match.params.id === '3'),
      );
      const customProps = { match: { params: { id: '3' } } };
      const authenticatedContainer = mount(<WithCustomFunction {...customProps} store={store} />);
      expect(authenticatedContainer).toMatchSnapshot();
    });
  });

  describe('user is not authorized', () => {
    const unauthorizedStore = configureStore()({
      entities: {
        users: {
          1: { id: '1', username: 'name' },
          2: { id: '2', username: 'name2' },
        },
      },
      signin: { currentUser: { users: ['2'] } },
    });

    it('renders Redirect component', () => {
      const unauthorizedContainer = mount((
        // keyLength={0} to stop snapshots from failing due to randomly generated keys
        <MemoryRouter keyLength={0} >
          <WrappedComponent
            {...defaultProps}
            store={unauthorizedStore}
          />
        </MemoryRouter>
      ));
      expect(unauthorizedContainer).toMatchSnapshot();
    });

    it('allows to set fallback location', () => {
      const WrappedComponentWithFallback = authorizeComponent(MockComponent, '/other_location');
      const unauthorizedContainer = mount((
        <MemoryRouter keyLength={0} >
          <WrappedComponentWithFallback
            {...defaultProps}
            store={unauthorizedStore}
          />
        </MemoryRouter>
      ));
      expect(unauthorizedContainer).toMatchSnapshot();
    });
  });
});
