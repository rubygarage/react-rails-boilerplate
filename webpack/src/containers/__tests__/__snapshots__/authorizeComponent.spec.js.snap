// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorizeComponent HoC user is authorized allows to use custom authorization function 1`] = `
<Connect(Component)
  match={
    Object {
      "params": Object {
        "id": "3",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component
    currentUser={
      Object {
        "id": "1",
        "username": "name",
      }
    }
    dispatch={[Function]}
    match={
      Object {
        "params": Object {
          "id": "3",
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <MockComponent
      currentUser={
        Object {
          "id": "1",
          "username": "name",
        }
      }
      dispatch={[Function]}
      match={
        Object {
          "params": Object {
            "id": "3",
          },
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
      storeSubscription={
        Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        }
      }
    >
      <div>
        Mock Component
      </div>
    </MockComponent>
  </Component>
</Connect(Component)>
`;

exports[`authorizeComponent HoC user is authorized renders given component 1`] = `
<Connect(Component)
  match={
    Object {
      "params": Object {
        "id": "1",
      },
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component
    currentUser={
      Object {
        "id": "1",
        "username": "name",
      }
    }
    dispatch={[Function]}
    match={
      Object {
        "params": Object {
          "id": "1",
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  >
    <MockComponent
      currentUser={
        Object {
          "id": "1",
          "username": "name",
        }
      }
      dispatch={[Function]}
      match={
        Object {
          "params": Object {
            "id": "1",
          },
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
      storeSubscription={
        Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        }
      }
    >
      <div>
        Mock Component
      </div>
    </MockComponent>
  </Component>
</Connect(Component)>
`;

exports[`authorizeComponent HoC user is not authorized allows to set fallback location 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/other_location",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/other_location",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(Component)
      match={
        Object {
          "params": Object {
            "id": "1",
          },
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Component
        currentUser={
          Object {
            "id": "2",
            "username": "name2",
          }
        }
        dispatch={[Function]}
        match={
          Object {
            "params": Object {
              "id": "1",
            },
          }
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <Redirect
          push={false}
          to="/other_location"
        />
      </Component>
    </Connect(Component)>
  </Router>
</MemoryRouter>
`;

exports[`authorizeComponent HoC user is not authorized renders Redirect component 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/404",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/404",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Connect(Component)
      match={
        Object {
          "params": Object {
            "id": "1",
          },
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Component
        currentUser={
          Object {
            "id": "2",
            "username": "name2",
          }
        }
        dispatch={[Function]}
        match={
          Object {
            "params": Object {
              "id": "1",
            },
          }
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
        storeSubscription={
          Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
        }
      >
        <Redirect
          push={false}
          to="/404"
        />
      </Component>
    </Connect(Component)>
  </Router>
</MemoryRouter>
`;
